### PR TITLE
feature/condo/DOMA-2461/notifications templates refactoring

### PR DESCRIPTION
**Pull request by AleX83Xpert**
This pull request contains the following things:
1. Deep refactoring of templates for messages (email, SMS, push, telegram)
2. New file structure for templates
3. Templates rendering including dynamic variables
4. Tests

### CODE REVIEW

File: apps/condo/domains/notification/templates.js [hunk 1]
```diff
@@ -1,278 +1,176 @@
...
+    }
 
-        return {
-            subject: 'Новая организация. (СББОЛ)',
-            text: `
-                Название: ${get(data, ['organization', 'name'])},
-                ИНН: ${get(data, ['organization', 'meta', 'inn'])},
-            `,
-        }
+    if (!Object.keys(MESSAGE_TRANSPORTS_RENDERERS).includes(transport)) {
```
Comment (akarjakin) [inline apps/condo/domains/notification/templates.js:154]: Please move `Object.keys(MESSAGE_TRANSPORTS_RENDERERS)` to static constant cause it contains cycles under the hood.

File: apps/condo/domains/notification/templates.js [hunk 2]
```diff
@@ -1,278 +1,176 @@
...
+ * @typedef {Object} MessageTemplateEnvironment
+ * @property {string} serverUrl - server url may use for building links
+ */
+
+/**
+ *
+ * @type {Object<string, function({message: Message, env: MessageTemplateEnvironment}): Object>}
+ */
+const MESSAGE_TRANSPORTS_RENDERERS = {
+    [SMS_TRANSPORT]: function ({ message, env }) {
```
Comment (akarjakin) [inline apps/condo/domains/notification/templates.js:111]: inlined functions are harder to read and understand, I'd move them out from object.

File: apps/condo/domains/notification/templates.test.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+                        // So, at least one of them must exist.
+                        const templateFileText = path.resolve(__dirname, folder, `${transport}.${DEFAULT_TEMPLATE_FILE_EXTENSION}`)
+                        const templateFileHtml = path.resolve(__dirname, folder, `${transport}.html.${DEFAULT_TEMPLATE_FILE_EXTENSION}`)
+                        if (!fs.existsSync(templateFileText) && !fs.existsSync(templateFileHtml)) {
+                            console.error(`No template file(s) for ${transport}: ${templateFileText} or ${templateFileHtml}, or single ${DEFAULT_TEMPLATE_FILE_NAME}`)
+                            result = false
+                        }
+                    } else {
+                        const templateFile = path.resolve(__dirname, folder, `${transport}.${DEFAULT_TEMPLATE_FILE_EXTENSION}`)
+                        const hasParticularTransportTemplate = fs.existsSync(templateFile)
```
Comment (akarjakin) [inline apps/condo/domains/notification/templates.test.js:87]: `hasTemplate` would be enough

File: apps/condo/lang/en/messages/MESSAGE_FORWARDED_TO_SUPPORT/default.njk [hunk 4]
```diff
@@ -0,0 +1,12 @@
+Mobile OS: {{ message.meta.os }}
```
Comment (akarjakin) [inline apps/condo/lang/en/messages/MESSAGE_FORWARDED_TO_SUPPORT/default.njk:1]: one more language to learn :( what is advantage here over JS function ?

File: packages/@core.keystone/errors.js [hunk 5]
```diff
@@ -124,6 +124,7 @@ class GQLError extends ApolloError {
         const message = template(fields.message)(fields.messageInterpolation)
         extensions.message = message
         if (context) {
+            // todo use i18n from apps/condo/domains/common/utils/localesLoader.js
```
Comment (akarjakin) [inline packages/@core.keystone/errors.js:127]: please add ticket for TODO
`// TODO(DOMA-xxxx): ...`

File: apps/condo/lang/ru/messages/CUSTOMER_IMPORTANT_NOTE_TYPE/default.njk [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Название: {{ message.meta.organization.name }},
```
Comment (akarjakin) [inline apps/condo/lang/ru/messages/CUSTOMER_IMPORTANT_NOTE_TYPE/default.njk:1]: so we have at least two sources of truth for translations...

File: apps/condo/domains/common/utils/localesLoader.js [hunk 7]
```diff
@@ -3,36 +3,60 @@ const fs = require('fs')
...
 
-const getTranslations = (lang = conf.DEFAULT_LOCALE) => {
+const maybeLoadTranslations = () => {
     if (isEmpty(translations)) {
         loadTranslations()
     }
+}
+
+const getTranslations = (lang = conf.DEFAULT_LOCALE) => {
+    maybeLoadTranslations()
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/common/utils/localesLoader.js:27]: Agree. Fixed.

File: apps/condo/domains/notification/constants/constants.js [hunk 8]
```diff
@@ -34,6 +34,17 @@ const MESSAGE_TYPES = [
     MESSAGE_FORWARDED_TO_SUPPORT,
 ]
 
+/**
+ * If some messages types has limited variety of transports, please set it here.
+ * The rest of types must have templates for all transports or at least default template.
+ */
+const MESSAGE_TYPES_TRANSPORTS = {
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/notification/constants/constants.js:41]: This tuning is exclusion. In general, all types of messages should have the ability to be sent by any transport. But in some rare cases, we need to use not all transports. For example, `MESSAGE_FORWARDED_TO_SUPPORT` is the internal message and can be sent only by email. So we do not need to set a title for push messages for this type.
On another side, we have tests that check all stuff for templating: files and translations. If we exclude push transport for some message type, so we do not have to set the title translation. And this setting prevents some translations from being mandatory.

File: apps/condo/domains/notification/templates.js [hunk 9]
```diff
@@ -1,278 +1,176 @@
...
-    CUSTOMER_IMPORTANT_NOTE_TYPE,
-    MESSAGE_FORWARDED_TO_SUPPORT,
+    TELEGRAM_TRANSPORT,
+    PUSH_TRANSPORT,
+    DEFAULT_TEMPLATE_FILE_NAME,
+    DEFAULT_TEMPLATE_FILE_EXTENSION,
 } = require('./constants/constants')
+const { i18n } = require('@condo/domains/common/utils/localesLoader')
+
+const langDirRelated = '../../lang'
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.js]: Fixed

File: apps/condo/domains/notification/templates.js [hunk 10]
```diff
@@ -1,278 +1,176 @@
...
+const templateEngineDefaultDateFormat = 'D MMMM YYYY'
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.js]: Fixed

File: apps/condo/domains/notification/templates.js (hunk 1 above)
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.js]: Done.

File: apps/condo/domains/notification/templates.js (hunk 2 above)
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.js]: Ok, done.

File: apps/condo/domains/notification/templates.test.js (hunk 3 above)
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.test.js]: Refactored

File: apps/condo/lang/en/messages/MESSAGE_FORWARDED_TO_SUPPORT/default.njk (hunk 4 above)
Comment (AleX83Xpert) [reply apps/condo/lang/en/messages/MESSAGE_FORWARDED_TO_SUPPORT/default.njk:1]: No, this is the [popular template engine](https://mozilla.github.io/nunjucks/templating.html). We are already using it for code generation (see `packages/@core.codegen`)

File: apps/condo/lang/ru/messages/CUSTOMER_IMPORTANT_NOTE_TYPE/default.njk (hunk 6 above)
Comment (AleX83Xpert) [reply apps/condo/lang/ru/messages/CUSTOMER_IMPORTANT_NOTE_TYPE/default.njk:1]: I didn't understand which two sources you mean.
`njk` templates are for complex messages including formatting.
`json` files are for translating simple strings.

File: packages/@core.keystone/errors.js (hunk 5 above)
Comment (AleX83Xpert) [reply packages/@core.keystone/errors.js:127]: This is not an urgent task. The next one who will edits this code will see this comment and maybe simplifies the code.

File: apps/condo/domains/common/utils/localesLoader.js [hunk 11]
```diff
@@ -1,38 +1,54 @@
...
-    }).reduce((prev, curr) => ({ ...prev, ...curr }))
+    const localeFolders = fs.readdirSync(translationsDir)
+    translations = localeFolders
+        .map(languageCode => ({
+            [languageCode]: require(path.join(translationsDir, `${languageCode}/${languageCode}.json`)),
+        }))
+        .reduce((prev, curr) => ({ ...prev, ...curr }))
 }
 
+loadTranslations()
```
Comment (pahaz) [inline apps/condo/domains/common/utils/localesLoader.js:19]: Why not to put it inside the _app?

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/common/utils/localesLoader.js (hunk 11 above)
Comment (AleX83Xpert) [reply apps/condo/domains/common/utils/localesLoader.js]: Not needed. This code was refactored.

File: apps/condo/domains/common/utils/localesLoader.js [hunk 12]
```diff
@@ -1,38 +1,49 @@
...
+ *   ...,
+ *   "greeting": "Hello, {name}!",
+ *   ...,
+ * }
+ *
+ * i18n('greeting', { meta: { name: 'World' } })
+ * // => "Hello, World!"
+ */
+const i18n = (code, { lang = conf.DEFAULT_LOCALE, meta = {} } = {}) => {
+    return template(get(translations, [lang, code], code), { interpolate: /{([\s\S]+?)}/g })(meta)
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/common/utils/localesLoader.js]: Done

File: apps/condo/domains/notification/templates.test.js [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+                    console.error(`There is no "${targetKey}" translation in ${locale}.json`)
+                    result = false
+                }
+            }
+        }
+
+        expect(result).toEqual(true)
+    })
+
+    it('All push templates has translations for titles', () => {
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.test.js]: fixed

File: apps/condo/domains/notification/templates.test.js [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+                        }
+                    }
+                }
+            }
+        }
+
+        expect(result).toEqual(true)
+    })
+
+    it('All email templates has subjects translations', () => {
```
[the remark this answers is not in the document]
Comment (AleX83Xpert) [reply apps/condo/domains/notification/templates.test.js]: fixed

**Review by akarjakin — APPROVED** [review]

**Pull request merged**